Feature / BJS2-52380 / Account Service / Balances

drewbal opened the pull request.

Commit:
Merge branch 'Dragon-master-stream7' into Dragon-master-stream7-Sprint-12

Commit:
Feature / Account Service / Adds Account Balance

Adds new migrations: change of index in Account table, change of Balance table primary key

Commit:
Feature / Account Service / Adds Account Balance

Adds new custom exceptions

Commit:
Feature / Account Service / Adds Account Balance

Adds AccountBalanceDto and TransactionDto

Commit:
Feature / Account Service / Adds Account Balance

Adds Account and Balance entities

Commit:
Feature / Account Service / Adds Account Balance

Adds methods to work with balance in Controller

Commit:
Feature / Account Service / Adds Account Balance

Changes the configuration of ObjectMapper: disables serialization of date as an array

Commit:
Feature / Account Service / Adds Account Balance

Adds balance management methods in Account Service

DeemaP commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+
+        BigDecimal transactionAmount = transactionDto.amount();
+
+        BigDecimal newAuthorizedAmount = account.getBalance().getAuthorizedBalance().subtract(transactionAmount);
```
в методах approve и cancel дублируются строки. Ты столько всего повыносил в приватные методы, что может и эти две строки вынести?)

drewbal replied:
Да, можно!

DeemaP reviewed (commented):
Я постарался на ревью кода, но не смог осилить плотное чтение тестов(((

DeemaP commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+    private final AccountEventPublisher accountEventPublisher;
+    private final AccountRepository accountRepository;
+
+    public AccountDto createAccount(CreateAccountDto dto, Long ownerId) {
```
Не нужна ли тут транзакция?

drewbal replied:
Руслан задал этот же вопрос в другом pr. Я поставил

Почему не ставил изначально? Этот метод вызывает только одну операцию базы данных, добавление новой строки, других операций нет.

DeemaP replied:
Смысл в том, что если у тебя не отработает паблишер, то в базе все равно сохранится запись, а зачем она нам нужна, если о ней не было события?

Поэтому логично использовать транзакцию, чтоб в случае, если редис поломался, мы не сохраняли запись в базу.

drewbal replied:
Это хорошо, но только Transactional не спасет. Он гарантирует выполнение только sql запросов, которые нагенерятся в этом методе через репозиторий. А сторонние факапы, будь то редис или s3 базу данных не волнуют, увы. Для этого мы и крутили то ретраи, то овербоксы

DeemaP replied:
Да, я забыл, что аутбокс юзал, а там добавление в базу :D

DeemaP commented on src/test/java/faang/school/accountservice/publisher/AccountEventPublisherTest.java:
```diff
@@ -0,0 +1,80 @@
+    @Test
+    @DisplayName("Account event published successfully")
+    void testPublish_ValidEvent_Success() {
+
```
отступ

DeemaP commented on src/test/java/faang/school/accountservice/publisher/AccountEventPublisherTest.java:
```diff
@@ -0,0 +1,80 @@
+    @Test
+    @DisplayName("Publish null event should not send to Redis")
+    void testPublish_NullEvent_NoSend() {
+
```
отступ

DeemaP commented on src/test/java/faang/school/accountservice/publisher/AccountEventPublisherTest.java:
```diff
@@ -0,0 +1,80 @@
+    private RedisTemplate<String, Object> redisTemplate;
+
+    private RedisProperties redisProperties;
+
```
здесь аннотаций нет, можно без отступов перечислить остальные поля.

drewbal replied:
окей, но я все равно оставляю пустую строку между компонентами и разными дто, примитивами и прочими

DeemaP commented on src/test/java/faang/school/accountservice/service/AccountServiceTest.java:
```diff
@@ -0,0 +1,369 @@
+                accountService.updateAccountStatus(accountNumber, ownerId, newStatus));
+
+        verify(accountRepository, times(1)).findByAccountNumber(accountNumber);
+        verify(accountRepository, times(0)).save(any(Account.class));
```
never() получше читается. И ниже есть еще times(0).

Еще недавно открыл для себя, на случай нескольки верифаев с never(), есть метод verifyNoInteractions() внутри которого можно просто перечислить объекты с которыми не было взаимодействий, но это применимо именно к объектам, а не методам.

drewbal replied:
Интересно. Я следую твоему совету и делаю только один never или times(0), т.к. после него уже не будет исполнения кода

DeemaP replied:
Это логично, да, но просто новый прикол открыл для себя, вот и поделился)
По сути тоже одна строка, но может заменить несколько строк verify(blablabla, never())

drewbal commented on src/main/java/faang/school/accountservice/config/AppConfig.java:
```diff
@@ -0,0 +1,19 @@
+    public ObjectMapper objectMapper() {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.registerModule(new JavaTimeModule());
+        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
```
Ха, это прикольная штука. По умолчанию objectMapper сериализует дату в массив типа [2024, 12, 26, 9, 34].

А если ее отключить, то тогда будет строка "2024-12-24T15:48:05.236815"

DeemaP replied:
О, вот это класс) Над в остальных сервисах прикрутить тоже)

drewbal commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -0,0 +1,116 @@
+    public ResponseEntity<AccountBalanceDto> deposit(@Valid @RequestBody TransactionDto transactionDto) {
+        log.info("Request to deposit funds on the account: number: {}, amount: {}", transactionDto.accountNumber(), transactionDto.amount());
+
+        return ResponseEntity.ok().body(accountService.deposit(transactionDto));
```
конечно, был невнимателен

drewbal commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -0,0 +1,116 @@
+
+    @PostMapping("/approve")
+    @Operation(summary = "Approve the pending transaction")
+    public ResponseEntity<Void> approve(@Valid @RequestBody TransactionDto transactionDto) {
```
Ты прав. Но валидация должна быть не по автору транзакции. Согласно логике подтверждение или отмена должны приходить от платежной системы, а не от простого юзера, т.е. это механизм клиринга. То есть тут скорее будет механизм скрытых от всех эндпойнтов, можно придумать через специальную версию api, например. И так же у нас должны быть отдельные сущности этих платежных агентов, но я это не делал. Запишу в бэклог

DeemaP replied:
Понял) Я подумал, что это процедура юзера, как доп подтверждение транзакции)
То есть ты как бы отправил денежку, а потом подтвердил или отменил) Условный вопрос "а ты уверен, что хочешь отправить деньги?"

drewbal commented on src/main/java/faang/school/accountservice/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @ExceptionHandler(MissingServletRequestParameterException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
```
Ты чертовски прав! 
Я просто скопипастил этот шаблон из прошлых методов, и не задумался. А ведь действительно избыточно, раз мы используем ResponseEntity. Надо в бэклог задачу по корректировке всех хэндлеров в других сервисах.

drewbal commented on src/main/java/faang/school/accountservice/config/RetryProperties.java:
```diff
@@ -0,0 +1,28 @@
+@ConfigurationProperties(prefix = "spring.retry")
+public class RetryProperties {
+
+    @Min(3)
```
На самом деле этот хардкод задумывался как проверка конфига. Типа если ты в конфиге поставишь неправильное значение, то приложение не запустится. Но увы, это хардкод, поэтому лучше его вообще убрать

drewbal commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+
+        log.info("Request to get accounts of {} id: {} completed", ownerType, ownerId);
+        return accountMapper.toDto(accounts);
+    }
```
ты имеешь ввиду между логом и return? Ну вот я думал об этом, получается столько пустых строк. А тут вроде как безобидный лог, заканчивающий операцию, и после него только return. Хз. Правда читаемость ухудшается?

DeemaP replied:
я имею в виду отступ между методами, у тебя слиплась аннотация и предыдущий метод

drewbal commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    public List<AccountDto> getAccounts(AccountOwnerType ownerType, Long ownerId) {
+        List<Account> accounts = accountRepository.findByOwnerTypeAndOwnerId(ownerType, ownerId);
```
Ну вот по идее мы это и делаем. Определенный пользователь может достать только свои счета, т.к. его id зашит в хэдере. Если видишь уязвимость, то подскажи

DeemaP replied:
отбой тревоги) Я упустил, что в эндпоинте у тебя присвоение userId идет к ownedId, сори)

drewbal commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+        }
+    }
+
+    private void validateDepositPossibility(AccountStatus status) {
```
Да. Пополнить можно любой счет, кроме закрытого. А вот снимать только при определенных статусах. Пример из жизни: налоговая может заблокировать счет, то есть вывести деньги нельзя будет. Но при этом на счет будут прилетать платежи, которые налоговая будет хватать себе.

DeemaP replied:
хм, принимается)

drewbal commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+        accountEventPublisher.publish(accountMapper.toDto(account));
+
+        log.info("Account status updated: number: {}, status: {}", account.getAccountNumber(), accountStatus);
+        return accountMapper.toDto(account);
```
И правда глупо. Спасибо

drewbal commented on src/main/java/faang/school/accountservice/service/AccountService.java:
```diff
@@ -0,0 +1,218 @@
+        }
+    }
+
+    private void validateWithdrawalPossibility(AccountStatus status) {
```
А можно попробовать, на самом деле

drewbal replied:
Более того, я изменил логику удаления счета и вообще убрал статус DELETED. Поэтому проверку делаем только на пополнение

DeemaP replied:
Ну по сути это относилось к предыдущему вопросу, так что не так актуально уже, но можно сделать один метод валидации, а внутри зашить логику в зависимости от статуса счета (хотя этот может оказаться усложнением)

drewbal replied:
Ты же код еще не видел :))

Commit:
Feature / Account Service / Adds Account Balance

Cross-review fixes

Commit:
Merge branch 'Dragon-Sprint-12-Account' into Dragon-Sprint12-balance

# Conflicts:
#	src/main/java/faang/school/accountservice/controller/AccountController.java
#	src/main/java/faang/school/accountservice/entity/Account.java
#	src/main/java/faang/school/accountservice/service/AccountService.java
#	src/test/java/faang/school/accountservice/controller/AccountControllerTest.java
#	src/test/java/faang/school/accountservice/service/AccountServiceTest.java

Commit:
Feature / Account Service / Adds Account Balance

Adds Transaction table

Commit:
Feature / Account Service / Adds Account Balance

Adds Transaction Entity, relevant enums and exceptions

Commit:
Feature / Account Service / Adds Account Balance

Adds the transactions functionality to the interactions with account and balance

Commit:
Feature / Account Service / Adds Account Balance

Minor fixes

Commit:
Merge branch 'Dragon-master-stream7' into Dragon-Sprint12-balance

Commit:
Feature / Account Service / Adds Account Balance

Minor fixes

RuslanZinovyev commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -79,4 +86,62 @@ public ResponseEntity<Void> deleteAccount(
+        return ResponseEntity.ok(accountService.withdraw(ownerId, transactionRequestDto));
+    }
+
+    @PostMapping("/{transactionId}/approve")
```
URL не очень понятен, по сути это `/accounts/12345/approve` - получается что id относится к аккаунту, а не транзакции.
Может вытащить из DTO  accountNumber?
`POST /accounts/{accountNumber}/transactions/{transactionId}/approve`

drewbal replied:
Я наоборот стремился к тому, чтобы непосредственно из URL убрать информацию по-максимуму. Данный эндпойнт будет пользоваться платежными системами для клиринга, поэтому оставил транзакции как PathVariable. Может лучше сделать /accounts/transactions/{transactionId}/approve?

Или accounts/transactions без {accountNumber} между ними плохо выглядит?

RuslanZinovyev commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -79,4 +86,62 @@ public ResponseEntity<Void> deleteAccount(
+        return ResponseEntity.ok().build();
+    }
+
+    @PostMapping("/{transactionId}/reject")
```
Тут тоже самое, `transactionId` идет после `/accounts`

drewbal replied:
аналогично комменту выше

RuslanZinovyev commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -79,4 +86,62 @@ public ResponseEntity<Void> deleteAccount(
         return ResponseEntity.noContent().build();
     }
+
+    @PostMapping("/deposit")
```
`/{accountNumber}/deposit`

drewbal replied:
Так изначально и было, но я решил, что номер аккаунта в пути это не очень хорошо. С другой стороны ДТО тоже не дает никакой приватности...

RuslanZinovyev commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -79,4 +86,62 @@ public ResponseEntity<Void> deleteAccount(
+        return ResponseEntity.ok(accountService.deposit(transactionRequestDto));
+    }
+
+    @PostMapping("/withdraw")
```
`/{accountNumber}/withdraw`

Commit:
Feature / Account Service / Adds Account Balance

Review fixes
